src/server/lock_manager/waiter_manager.rs

```diff
@@ -242,69 +439,71 @@ impl WaiterManager {
+            .add_waiter(waiter)
+            .and_then(|old| {
+                old.notify();
+                Some(())
+            });
+        handle.spawn(f);
     }
 
-    fn handle_wake_up(
-        &mut self,
-        handle: &Handle,
-        lock_ts: TimeStamp,
-        hashes: Vec<u64>,
-        commit_ts: TimeStamp,
-    ) {
-        let mut ready_waiters = self
-            .wait_table
-            .borrow_mut()
-            .take_ready_waiters(lock_ts, hashes);
-        ready_waiters.sort_unstable_by_key(|waiter| waiter.start_ts);
-
-        for (i, waiter) in ready_waiters.into_iter().enumerate() {
-            self.detector_scheduler
-                .clean_up_wait_for(waiter.start_ts, waiter.lock);
-            if self.wake_up_delay_duration > 0 {
-                // Sleep a little so the transaction with small start_ts will more likely get the lock.
-                let when = Instant::now()
-                    + Duration::from_millis(self.wake_up_delay_duration * (i as u64));
-                let timer = Delay::new(when)
-                    .and_then(move |_| {
-                        wake_up_waiter(waiter, commit_ts);
-                        Ok(())
-                    })
-                    .map_err(|e| info!("wake-up timer delay errored"; "err" => ?e));
-                handle.spawn(timer);
-            } else {
-                wake_up_waiter(waiter, commit_ts);
+    fn handle_wake_up(&mut self, lock_ts: TimeStamp, hashes: Vec<u64>, commit_ts: TimeStamp) {
+        let mut wait_table = self.wait_table.borrow_mut();
+        let (mut visited, initial_len) = (0, wait_table.len());
+        let new_timeout = Instant::now() + Duration::from_millis(self.wake_up_delay_duration);
+        for hash in hashes {
+            let lock = Lock { ts: lock_ts, hash };
+            if let Some((mut oldest, others)) = wait_table.remove_oldest_waiter(lock) {
+                // Notify the oldest one immediately.
+                self.detector_scheduler
+                    .clean_up_wait_for(oldest.start_ts, oldest.lock);
+                oldest.conflict_with(lock_ts, commit_ts);
+                oldest.notify();
+                // Others will be waked up after `wake_up_delay_duration`.
+                //
+                // NOTE: Actually these waiters are waiting for an unknown transaction.
+                // If there is a deadlock between them, it will be detected after timeout.
+                if others.is_empty() {
+                    // Remove the empty entry here.
+                    wait_table.remove(lock);
+                } else {
+                    others.iter_mut().for_each(|waiter| {
+                        waiter.conflict_with(lock_ts, commit_ts);
+                        waiter.reset_timeout(new_timeout);
+                    });
+                }
+                // Break if we haved visited all entries to reduce the cost of small wait table
+                // against big hashes.
+                visited += 1;
```

Comment:
I think this optimization does not make much difference.
It benefits only if all the waiters are waiting for one transaction.
And in this case, only a small portion of keys saves the lookup.

Reply:
Yes. But it can't get worse.

Reply:
Okay. I will remove it.